### DISCUSSION TITLE
large JPEG 2000 file conversion takes a lot of memory

### DISCUSSION THREAD

**Discussion by MichaelTran262** [libvips/libvips#2383] [Q&A] [answered]
Hello,

I am using libvips 8.11.2 for jp2 conversion to jpg, png and tiff using `jp2kload`. When I convert large ~750 MB jp2 file, libvips uses approximately more than 6 GB of RAM. Is it possible to convert large JPEG2000 files without taking this large amounts of memory? I wish to deploy libvips into a docker container with 4 GB memory limit to convert and download image files. 

Thanks in advance.

**Comment by jcupitt** [comment]
Hi @michaeltran262,

My guess would be you're not using the libvips jp2k loader and it's fallen back to the imagemagick jp2k loader, which does use a lot of memory.

For example, with a 10,000 x 10,000 pixel jp2 image and the imagemagick loader, I see:

```
$ /usr/bin/time -f %M:%e vips copy wtc.jp2 x.jpg
1276272:11.96
```

1.2GB of memory and 10s of runtime. If I enable the native libvips jp2k loader, I see:

```
$ /usr/bin/time -f %M:%e vips copy wtc.jp2 x.jpg
147640:4.61
```

150mb and 5s of runtime.

You can see which loader it is using with `vipsheader`:

```
$ vipsheader wtc.jp2
wtc.jp2: 9372x9372 uchar, 3 bands, srgb, jp2kload
```

That's the native libvips loader. If it's using imagemagick, you'll see `magickload`.

**Reply by MichaelTran262** [reply]
This is really odd, I should be using jp2k loader.
```
$ vipsheader map.jp2
map.jp2: 26348x19036 uchar, 3 bands, srgb, jp2kload
```
When I try to copy map.jp2 the process will use over 7 GB of memory.

```
$ /usr/bin/time -f %M:%e vips copy map.jp2 x.jpg
7185096:55.86
```
The same process would be always killed in my docker container.

I compiled and installed OpenJPEG 2.4.0 from it's latest release source code (Ubuntu repository has older version of `libopenjp2-7`)
```
$ pkg-config libopenjp2 --print-provides
libopenjp2 = 2.4.0
```
and then I installed libvips with my desired configuration. Other libraries like `libjpeg-turbo8 libtiff-dev libpng-dev` were installed using apt from Ubuntu repository (Debian in case of Docker container).

Any help would be very appreciated.

**Comment by jcupitt** [comment]
Huh that's odd. Could the jp2 image have been written with very large tiles?

I see:

```
$ vips copy CMU-2.svs x.jp2
$ ls -l x.jp2
-rw-r--r-- 1 john john 786456862 Aug  6 17:57 x.jp2
$ /usr/bin/time -f %M:%e vips copy x.jp2 x.png
1129068:267.41
```

So a 800mb jp2 reads with 1.1gb of memory.

**Reply by MichaelTran262** [reply]
How can I see if an image was written with large tiles and how can I change it so it doesn't take more than it should (<4 GB)? 

But if your 800 MB jp2 file uses only 1.1gb of memory, then I think the problem is in my installation process. I already reinstalled many times and the conversion for my image still takes more than 6 GB, even for `opj_decompress`.

Here is my Dockerfile.

```
ARG OPENJPEG_VERSION=2.4.0
ARG OPENJPEG_URL=https://github.com/uclouvain/openjpeg/archive
ARG LIBVIPS_VERSION=8.11.2
ARG LIBVIPS_URL=https://github.com/libvips/libvips/releases/download

RUN apt-get update
RUN apt-get install -y \
	build-essential \
	cmake  \
	autoconf \
	pkg-config \
	glib2.0-dev \
        libexpat-dev \
	libjpeg62-turbo-dev \
        libpng-dev \
	libtiff-dev \
	liblcms2-dev \
	libz-dev \
	libexif-dev \
	time

# OpenJPEG 2.4.0
RUN cd /root \
	&& wget ${OPENJPEG_URL}/v${OPENJPEG_VERSION}.tar.gz \
	&& tar xf v${OPENJPEG_VERSION}.tar.gz \ 
	&& rm v${OPENJPEG_VERSION}.tar.gz  \
	&& cd openjpeg-${OPENJPEG_VERSION} \ 
	&& mkdir -v build \
	&& cd build \
	&& cmake .. -DCMAKE_BUILD_TYPE=Release \
	&& make \
	&& make install

# Libvips 8.11.2
RUN cd /root \
	&& wget ${LIBVIPS_URL}/v${LIBVIPS_VERSION}/vips-${LIBVIPS_VERSION}.tar.gz \
	&& tar xf vips-${LIBVIPS_VERSION}.tar.gz \
	&& rm vips-${LIBVIPS_VERSION}.tar.gz \
	&& cd vips-${LIBVIPS_VERSION} \
	&& ./configure \
	&& make V=0 \
	&& make install \
	&& ldconfig
```
I took and modified your Dockerfile from your docker-builds. I can't find anything wrong.

**Comment by jcupitt** [comment, marked as answer]
Have you tried creating a large jp2 file with libvips? It creates 512 x 512 tiles by default, but you can set any size.

If files written by libvips load in a reasonable amount of RAM, then yes, the issue is probably that your original images have huge tiles.

If you enable debugging in jp2kload.c (just uncomment the `#define DEBUG`) it'll print out some useful information about the image structure, eg.:

```
$ vipsheader x.jp2
vips_foreign_load_jp2k_build:
vips_foreign_load_jp2k_header:
image:
x0 = 0, y0 = 0, x1 = 78000, y1 = 30462, numcomps = 4, color_space = 0
icc_profile_buf = (nil), icc_profile_len = 0
components:
0) dx = 1, dy = 1, w = 78000, h = 30462, x0 = 0, y0 = 0
    prec = 8, bpp = 0, sgnd = 0, resno_decoded = 0, factor = 0
    data = (nil), alpha = 0
1) dx = 1, dy = 1, w = 78000, h = 30462, x0 = 0, y0 = 0
    prec = 8, bpp = 0, sgnd = 0, resno_decoded = 0, factor = 0
    data = (nil), alpha = 0
2) dx = 1, dy = 1, w = 78000, h = 30462, x0 = 0, y0 = 0
    prec = 8, bpp = 0, sgnd = 0, resno_decoded = 0, factor = 0
    data = (nil), alpha = 0
3) dx = 1, dy = 1, w = 78000, h = 30462, x0 = 0, y0 = 0
    prec = 8, bpp = 0, sgnd = 0, resno_decoded = 0, factor = 0
    data = (nil), alpha = 0
info:
tx0 = 0, ty0 = 0, tdx = 512, tdy = 512, tw = 153, th = 60
nbcomps = 4, tile_info = (nil)
x.jp2: 78000x30462 uchar, 4 bands, srgb, jp2kload
vips_foreign_load_jp2k_dispose:
```

So four image components, each 78k x 30k pixels, each tile is 512 x 512 pixels, and there are 153 x 60 of them.

**Reply by MichaelTran262** [reply]
I see now. My image has just one large tile. 
```
$ vipsheader map.jp2
vips_foreign_load_jp2k_build:
vips_foreign_load_jp2k_header:
image:
x0 = 0, y0 = 0, x1 = 26348, y1 = 19036, numcomps = 3, color_space = 0
icc_profile_buf = (nil), icc_profile_len = 0
components:
0) dx = 1, dy = 1, w = 26348, h = 19036, x0 = 0, y0 = 0
    prec = 8, bpp = 0, sgnd = 0, resno_decoded = 0, factor = 0
    data = (nil), alpha = 0
1) dx = 1, dy = 1, w = 26348, h = 19036, x0 = 0, y0 = 0
    prec = 8, bpp = 0, sgnd = 0, resno_decoded = 0, factor = 0
    data = (nil), alpha = 0
2) dx = 1, dy = 1, w = 26348, h = 19036, x0 = 0, y0 = 0
    prec = 8, bpp = 0, sgnd = 0, resno_decoded = 0, factor = 0
    data = (nil), alpha = 0
info:
tx0 = 0, ty0 = 0, tdx = 26348, tdy = 19036, tw = 1, th = 1
nbcomps = 3, tile_info = (nil)
map.jp2: 26348x19036 uchar, 3 bands, srgb, jp2kload
vips_foreign_load_jp2k_dispose:

```
Is there any way to avoid such high memory consumption with this large tile?

**Reply by jcupitt** [reply]
No, sorry. OpenJPEG will always allocate several tile-sized buffers for decode, so tile size sets memory use. You'll need to recode your images if you want to lower memory use.

**Reply by MichaelTran262** [reply]
That is unfortunate. But thank you nonetheless! I learned a lot.
